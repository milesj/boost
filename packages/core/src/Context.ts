import { isObject } from '@boost/common';

export default class Context {
  /**
   * Create a new instance of the current context and shallow clone all properties.
   */
  clone(...args: any[]): this {
    // @ts-expect-error
    const context = new this.constructor(...args);

    // Copy enumerable properties
    Object.keys(this).forEach(key => {
      const prop = key as keyof this;
      let value: any = this[prop];

      if (Array.isArray(value)) {
        value = [...value];
      } else if (value instanceof Map) {
        value = new Map(value);
      } else if (value instanceof Set) {
        value = new Set(value);
      } else if (value instanceof Date) {
        value = new Date(value.getTime());
      } else if (isObject<object>(value)) {
        // Dont dereference instances, only plain objects
        if (value.constructor === Object) {
          value = { ...value };
        }
      }

      context[prop] = value;
    });

    return context;
  }
}
