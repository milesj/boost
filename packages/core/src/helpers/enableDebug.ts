/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import debug from 'debug';

export default function enableDebug(namespace: string) {
  const { DEBUG } = process.env;
  const flag = `${namespace}:*`;

  if (DEBUG) {
    if (DEBUG.includes(flag)) {
      return;
    }

    process.env.DEBUG += `,${flag}`;
  } else {
    process.env.DEBUG = flag;
  }

  debug.enabled(flag);
}
