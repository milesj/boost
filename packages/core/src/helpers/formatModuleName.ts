/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

/**
 * Combine app and plugin names into a Node/NPM applicable module name.
 */
export default function formatModuleName(
  appName: string,
  pluginType: string,
  name: string,
  scoped: boolean = false,
): string {
  const moduleName = name.toLowerCase().replace(`${pluginType}:`, '');

  if (scoped) {
    return `@${appName}/${pluginType}-${moduleName}`;
  }

  return `${appName}-${pluginType}-${moduleName}`;
}
