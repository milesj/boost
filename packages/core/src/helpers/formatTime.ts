import prettyMs from 'pretty-ms';

export default function formatTime(ms: number): string {
  if (!Number.isFinite(ms) || ms === 0) {
    return '0s';
  }

  return prettyMs(ms, { keepDecimalsOnWholeSeconds: true });
}
