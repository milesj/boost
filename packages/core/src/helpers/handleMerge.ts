/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

/**
 * Handle special cases when merging 2 configuration values.
 * If the target and source are both arrays, concatenate them.
 */
export default function handleMerge(target: any, source: any): any {
  if (Array.isArray(target) && Array.isArray(source)) {
    return Array.from(new Set([...target, ...source]));
  }

  // Defer to lodash
  return undefined;
}
