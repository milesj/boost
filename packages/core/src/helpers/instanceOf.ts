/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Constructor } from '../types';

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and built files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export default function instanceOf<T = any>(object: any, contract: Constructor<T>): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === contract.name) {
      return true;
    }

    current = Object.getPrototypeOf(current);
  }

  return false;
}
