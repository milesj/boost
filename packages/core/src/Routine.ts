import execa, { Options as ExecaOptions, ExecaChildProcess, ExecaReturns } from 'execa';
import split from 'split';
import { Readable } from 'stream';
import optimal, { predicates, Blueprint, Predicates } from 'optimal';
import { Event } from '@boost/event';
import Context from './Context';
import Task, { TaskAction } from './Task';
import CoreTool from './Tool';
import { AggregatedResponse } from './Executor';
import ParallelExecutor from './executors/Parallel';
import PoolExecutor, { PoolExecutorOptions } from './executors/Pool';
import SerialExecutor from './executors/Serial';
import SyncExecutor from './executors/Sync';
import instanceOf from './helpers/instanceOf';
import wrapWithPromise from './helpers/wrapWithPromise';
import { STATUS_RUNNING } from './constants';
import { Debugger } from './types';

export interface CommandOptions {
  shell?: boolean;
  task?: Task<any>;
  wrap?: (process: ExecaChildProcess) => void;
}

export default abstract class Routine<
  Ctx extends Context,
  Tool extends CoreTool<any>,
  Options extends object = {}
> extends Task<Ctx> {
  // @ts-ignore Set after instantiation
  debug: Debugger;

  key: string = '';

  onCommand: Event<[string]>;

  onCommandData: Event<[string, string]>;

  options: Options;

  parent: Routine<Ctx, Tool> | null = null;

  // TODO Change to Set in 2.0
  routines: Routine<Ctx, Tool>[] = [];

  // TODO Change to Set in 2.0
  tasks: Task<Ctx>[] = [];

  // @ts-ignore Set after instantiation
  tool: Tool;

  constructor(key: string, title: string, options: Partial<Options> = {}) {
    super(title, (context, value) => this.execute(context, value));

    if (!key || typeof key !== 'string') {
      throw new Error('Routine key must be a valid unique string.');
    }

    this.key = key;
    this.options = optimal(options, this.blueprint(predicates), {
      name: this.constructor.name,
    });

    this.onCommand = new Event('command');
    this.onCommandData = new Event('command.data');
  }

  /**
   * Define an optimal blueprint in which to validate and build the
   * options passed to the constructor.
   */
  blueprint(preds: Predicates): Blueprint<Options> {
    return {} as any;
  }

  /**
   * Called once the routine has been configured and is ready to execute.
   */
  bootstrap() {}

  /**
   * Configure the routine after it has been instantiated.
   */
  configure(parent: Routine<Ctx, Tool>): this {
    this.parent = parent;
    this.tool = parent.tool;
    this.context = parent.context;
    this.metadata.depth = parent.metadata.depth + 1;

    // Custom debugger for this routine
    this.debug = this.tool.createDebugger('routine', this.key);
    this.debug('Bootstrapping routine');

    // Initialize routine
    this.bootstrap();

    return this;
  }

  /**
   * Execute a command with the given arguments and pass the results through a promise.
   */
  async executeCommand(
    command: string,
    args: string[],
    options: ExecaOptions & CommandOptions = {},
  ): Promise<ExecaReturns> {
    const { shell, task, wrap, ...opts } = options;
    const stream = shell
      ? execa.shell(`${command} ${args.join(' ')}`.trim(), opts)
      : execa(command, args, opts);

    this.onCommand.emit([command]);

    // Push chunks to the reporter
    const unit = task || this;
    const handler = (line: string) => {
      if (unit.status === STATUS_RUNNING) {
        unit.output += line;

        // Only capture the status when not empty
        if (line) {
          unit.statusText = line;
        }

        this.onCommandData.emit([command, line]);
      }
    };

    if (!shell) {
      (stream.stdout as Readable).pipe(split()).on('data', handler);
      (stream.stderr as Readable).pipe(split()).on('data', handler);
    }

    // Allow consumer to wrap functionality
    if (typeof wrap === 'function') {
      wrap(stream as ExecaChildProcess);
    }

    return wrapWithPromise(stream);
  }

  /**
   * Execute routines in parallel.
   */
  async parallelizeRoutines<T>(value?: T, routines?: Routine<Ctx, Tool>[]): Promise<any[]> {
    return new ParallelExecutor(this.tool, this.context).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in parallel.
   */
  async parallelizeTasks<T>(value?: T, tasks?: Task<Ctx>[]): Promise<any[]> {
    return new ParallelExecutor(this.tool, this.context).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Add a new routine within this routine.
   */
  pipe(routine: Routine<Ctx, Tool>): this {
    if (instanceOf(routine, Routine)) {
      // eslint-disable-next-line no-param-reassign
      routine.metadata.index = this.routines.length;

      this.routines.push(routine.configure(this));
    } else {
      throw new TypeError(this.tool.msg('errors:routineInstanceInvalid'));
    }

    return this;
  }

  /**
   * Execute routines in a pool.
   */
  async poolRoutines<T>(
    value?: T,
    options?: Partial<PoolExecutorOptions>,
    routines?: Routine<Ctx, Tool>[],
  ): Promise<AggregatedResponse> {
    return new PoolExecutor(this.tool, this.context, options).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in a pool.
   */
  async poolTasks<T>(
    value?: T,
    options?: Partial<PoolExecutorOptions>,
    tasks?: Task<Ctx>[],
  ): Promise<AggregatedResponse> {
    return new PoolExecutor(this.tool, this.context, options).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Execute routines in sequential (serial) order.
   */
  serializeRoutines<T>(value?: T, routines?: Routine<Ctx, Tool>[]): Promise<any> {
    return new SerialExecutor(this.tool, this.context).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in sequential (serial) order.
   */
  serializeTasks<T>(value?: T, tasks?: Task<Ctx>[]): Promise<any> {
    return new SerialExecutor(this.tool, this.context).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Execute routines in sync.
   */
  async synchronizeRoutines<T>(
    value?: T,
    routines?: Routine<Ctx, Tool>[],
  ): Promise<AggregatedResponse> {
    return new SyncExecutor(this.tool, this.context).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in sync.
   */
  async synchronizeTasks<T>(value?: T, tasks?: Task<Ctx>[]): Promise<AggregatedResponse> {
    return new SyncExecutor(this.tool, this.context).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Define an individual task.
   */
  task(title: string, action: TaskAction<Ctx>, scope?: any): Task<Ctx> {
    if (typeof action !== 'function') {
      throw new TypeError(this.tool.msg('errors:taskRequireAction'));
    }

    const task = new Task<Ctx>(title, action.bind(scope || this));

    task.parent = this;
    task.metadata.index = this.tasks.length;

    this.tasks.push(task);

    return task;
  }

  /**
   * Execute the current routine and return a new value.
   * This method *must* be overridden in a subclass.
   */
  abstract async execute(context: Ctx, value: any): Promise<any>;
}
