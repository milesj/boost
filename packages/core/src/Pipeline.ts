import exit from 'exit';
import path from 'path';
import util from 'util';
import { Blueprint, Predicates } from 'optimal';
import { instanceOf } from '@boost/common';
import { CrashReporter } from '@boost/debug';
import Context from './Context';
import ExitError from './ExitError';
import Routine from './Routine';
import CoreTool from './Tool';

export interface PipelineOptions {
  exit?: (code: number) => void;
}

export default class Pipeline<Ctx extends Context, Tool extends CoreTool<any>> extends Routine<
  Ctx,
  Tool,
  PipelineOptions
> {
  constructor(tool: Tool, context: Ctx, options?: PipelineOptions) {
    super('root', 'Pipeline', options);

    if (instanceOf(tool, CoreTool)) {
      tool.initialize();
    } else {
      throw new TypeError('A `Tool` instance is required to operate the pipeline.');
    }

    this.tool = tool;
    this.tool.debug('Instantiating pipeline');

    // Child routines should start at 0
    this.metadata.depth = -1;

    this.setContext(context);
  }

  blueprint({ func }: Predicates): Blueprint<Required<PipelineOptions>> {
    return {
      exit: func(exit).notNullable(),
    };
  }

  /**
   * Execute all routines in order.
   */
  run<T>(initialValue?: T): Promise<any> {
    const { console: cli } = this.tool;

    this.tool.debug('Running pipeline');

    cli.start([Array.from(this.routines), initialValue]);

    return this.serializeRoutines(initialValue)
      .then(result => {
        cli.stop();

        process.exitCode = 0;

        return result;
      })
      .catch(error => {
        cli.stop(error);

        this.reportCrash(error);

        if (instanceOf(error, ExitError)) {
          this.options.exit(error.code);
        } else {
          this.options.exit(1);
        }

        return error;
      });
  }

  /**
   * Report the pipeline failure by writing a crash log.
   */
  reportCrash(error: Error) {
    const { config, options } = this.tool;
    const reporter = new CrashReporter()
      .reportBinaries()
      .reportProcess()
      .reportSystem();

    reporter
      .addSection('Tool Instance')
      .add('App name', options.appName)
      .add('App path', options.appPath)
      .add('Plugin types', Object.keys(this.tool.getRegisteredPlugins()).join(', '))
      .add('Scoped package', options.scoped ? 'Yes' : 'No')
      .add('Root', options.root)
      .add('Config name', options.configName)
      .add('Package path', path.join(options.root, 'package.json'))
      .add('Workspaces root', options.workspaceRoot || '(Not enabled)')
      .add(
        'Extending configs',
        config.extends.length > 0 ? util.inspect(config.extends) : '(Not extending)',
      );

    reporter
      .reportLanguages()
      .reportStackTrace(error)
      .reportEnvVars()
      .write(path.join(options.root, `${options.appName}-error.log`));
  }
}
