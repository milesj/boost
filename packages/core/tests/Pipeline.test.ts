import Pipeline from '../src/Pipeline';
import Routine from '../src/Routine';
import { createTestTool, createTestRoutine } from './helpers';
import Context from '../src/Context';

describe('Pipeline', () => {
  let pipeline: Pipeline<any>;

  beforeEach(() => {
    pipeline = new Pipeline(createTestTool(), new Context());
  });

  describe('constructor()', () => {
    it('errors if no tool is passed', () => {
      // @ts-ignore
      expect(() => new Pipeline()).toThrowErrorMatchingSnapshot();
    });

    it('sets the context', () => {
      pipeline = new Pipeline(createTestTool(), { foo: 'bar' });

      expect(pipeline.context).toEqual({ foo: 'bar' });
    });
  });

  describe('run()', () => {
    it('starts console with routine', async () => {
      const spy = jest.fn();
      const routine = createTestRoutine(pipeline.tool);

      pipeline.tool.console.emit = spy;
      pipeline.pipe(routine);

      await pipeline.run();

      expect(spy).toHaveBeenCalledWith('start', [[routine]]);
    });

    it('exits the console on success', async () => {
      const spy = jest.fn();

      pipeline.tool.console.exit = spy;

      await pipeline.run();

      expect(spy).toHaveBeenCalledWith(null, 0);
    });

    it('exits the console on failure', async () => {
      class FailureRoutine extends Routine<any, any> {
        execute() {
          return Promise.reject(new Error('Oops'));
        }
      }

      const spy = jest.fn();

      pipeline.tool.console.exit = spy;

      try {
        await pipeline.pipe(new FailureRoutine('fail', 'title')).run();
      } catch (error) {
        expect(error).toEqual(new Error('Oops'));
      }

      expect(spy).toHaveBeenCalledWith(new Error('Oops'), 1);
    });
  });
});
