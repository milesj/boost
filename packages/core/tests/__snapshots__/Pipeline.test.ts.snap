// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline constructor() errors if no tool is passed 1`] = `"A build \`Tool\` instance is required to operate the pipeline."`;
